Fix issue #1925 - UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow

AlonEirew opened the pull request:
**Proposed changes**:
- Convert the list of nparray to a python list

**Status (please check what you already did)**:
- [X] First draft (up for discussions & feedback)
- [ ] Final code
- [ ] Added tests
- [ ] Updated documentation

closes #1925

julian-risch commented:
Hi @AlonEirew I saw that the error message you reported in the issue recommends: "Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor.".
Have you tried that? Did you find any specific reason to use `to_list()` and converted the nparray to a python list instead?

AlonEirew commented:
Hi @julian-risch, I preferred this solution as all lists in the dictionary `sample` are python lists (with the exception of `ctx_segment_ids`), so I felt it will be more homogeneous and simple. I only tested this method, though I'm sure the method suggested by the error message will work as well.

julian-risch commented:
Hi @AlonEirew I am wondering whether the two options (or actually the three options including no conversion at all) differ in speed. Would it be possible for you to find that out please? The warning message and the reason for this PR is about speed so that's why I would like to make sure that we are actually improving speed with this PR.

AlonEirew commented:
Hi @julian-risch I've made the following changes to the branch:
1) I've removed the sharing strategy which I've accidentally merged to this branch (and belong to a separate PR).
2) I've changed the numpy.zeros_like method with a python list

Below code and measurement comparing the two methods, the original zeros_like method (before any of my changes) vs. new method of using python list:

**Code**
```
import timeit
starttime1 = timeit.default_timer()
# Original code (before any changes)
ctx_segment_ids1 = np.zeros_like(ctx_inputs["token_type_ids"], dtype=np.int32)
endtime1 = timeit.default_timer()
print("zeros_like time:", endtime1 - starttime1)

# New code
starttime2 = timeit.default_timer()
ctx_segment_ids2 = [[0] * len(ctx_inputs["token_type_ids"][0])] * len(ctx_inputs["token_type_ids"])
endtime2 = timeit.default_timer()
print("python_list time:", endtime2 - starttime2)
```
**Output**
```
zeros_like time: 0.00035607069730758667
python_list time: 2.222508192062378e-05
```
This method is 10x faster, and also removes the UserWarning

julian-risch commented:
I agree that having `ctx_segment_ids` as a list makes `sample` more homogeneous and simple. Further, it's plausible that `np.zeros_like` is slower than the changed code used in this PR. So, let's merge this. 👍
